### DISCUSSION TITLE
Seems DistinctByIterator much clearer than DistinctIterator, any chance to make DistinctIterator simpler?

### DISCUSSION THREAD

**Discussion by jjshao-ms** [dotnet/runtime#90514] [Ideas] [answered]
The DistinctByIterator's implementation seems to be much clearer than
https://github.com/dotnet/runtime/blob/71a3d36e5cdaf1e2d41957fb1d2cebff8af0e063/src/libraries/System.Linq/src/System/Linq/Distinct.cs#L62-L79
the DistinctIterator one
https://github.com/dotnet/runtime/blob/71a3d36e5cdaf1e2d41957fb1d2cebff8af0e063/src/libraries/System.Linq/src/System/Linq/Distinct.cs#L85-L150

Any chance to make DistinctIterator looks like DistinctByIterator? Or is there any performance/code size related background prevent us from doing so?

**Comment by stephentoub** [comment, marked as answer]
DistinctIterator is written that way because it allows implementing additional interfaces on the type, something Distinct can take advantage of more so than DistinctBy. You can see how that's utilized in the Distinct.SpeedOpt.cs file.
